File: data_structures/Backtracking/sudoku.cpp
/*

Input:
1
3 0 6 5 0 8 4 0 0
5 2 0 0 0 0 0 0 0
0 8 7 0 0 0 0 3 1
0 0 3 0 1 0 0 8 0
9 0 0 8 6 3 0 0 5
0 5 0 0 9 0 6 0 0
1 3 0 0 0 0 2 5 0
0 0 0 0 0 0 0 7 4
0 0 5 2 0 6 3 0 0

Output:
3 1 6 5 7 8 4 9 2
5 2 9 1 3 4 7 6 8
4 8 7 6 2 9 5 3 1
2 6 3 4 1 5 9 8 7
9 7 4 8 6 3 1 2 5
8 5 1 7 9 2 6 4 3
1 3 8 9 4 7 2 5 6
6 9 2 3 5 1 8 7 4
7 4 5 2 8 6 3 1 9

*/

#include<iostream>
using namespace std;
#define n 9

void print(int a[n][n])
{
    int i,j;
    
    for(i=0;i<9;i++)
       {
           for(j=0;j<9;j++)
            cout<< a[i][j]<<" ";
       } 
       
       cout<<"\n";
}

bool rowsafe(int a[n][n], int row, int num )
{
    int i,j;
    
    for(i=0;i<9;i++)
        if(a[row][i]==num)
            return false;
            
    return true;
}



bool colsafe(int a[n][n], int col, int num )
{
    int i,j;
    
    for(i=0;i<9;i++)
        if(a[i][col]==num)
            return false;
            
    return true;
}



bool boxsafe(int a[n][n], int row,int col, int num )
{
    int i,j;
    
    for(i=0;i<3;i++)
        for(j=0;j<3;j++)
            if(a[i+row][j+col]==num)
                return false;
            
    return true;
}


bool safe(int a[n][n], int row, int col, int num)
{
    if(rowsafe(a,row,num) && colsafe(a,col,num) && boxsafe(a,row-row%3,col-col%3,num) && a[row][col]==0)
        return true;
    
    return false;
}
bool notassigned(int a[n][n], int &r, int &c)
{
    int i,j;
    
    for(r=0;r<n;r++)
        for(c=0;c<n;c++)
            if(a[r][c]==0)
                return false;
                
    return true;
}

bool solve(int a[n][n])
{
    int r, c;
    
    if(notassigned(a, r, c))
        return true;
        
    
    int i,j;
    
    for(i=1;i<=9;i++)
    {
        if(safe(a,r,c,i))
        {
            a[r][c]=i;
            if(solve(a))
                return true;
            a[r][c]=0;
        }
    }
    return false;
    
    
}

int main()
 {
	//code
	
	int t;
	cin>>t;
	while(t--)
	{
	    int a[n][n];
	    
	    int i,j;
	    
	    for(i=0;i<n;i++)
	        for(j=0;j<n;j++)
	            cin>>a[i][j];
	    
	    
	   if(solve(a))
	        print(a);
	    
	   else
	    cout<<-1<<"\n";
	    
	}
	
	return 0;
}

File: sudoku.py
sudoku = [
    [7,8,0,4,0,0,1,2,0],
    [6,0,0,0,7,5,0,0,9],
    [0,0,0,6,0,1,0,7,8],
    [0,0,7,0,4,0,2,6,0],
    [0,0,1,0,5,0,9,3,0],
    [9,0,4,0,6,0,0,0,5],
    [0,7,0,3,0,0,0,1,2],
    [1,2,0,0,0,7,4,0,0],
    [0,4,9,2,0,6,0,0,7]
]

def solve(a):
    b = find(a)
    if not b:
        return True
    else:
        row, col = b
    for i in range(1,10):
        if check(a, i, (row, col)):
            a[row][col] = i
            if solve(a):
                return True
            a[row][col] = 0
    return False

def check(a, num, pos):
    # Check row
    for i in range(len(a[0])):
        if a[pos[0]][i] == num and pos[1] != i:
            return False
    for i in range(len(a)):
        if a[i][pos[1]] == num and pos[0] != i:
            return False
    x = pos[1] // 3
    y = pos[0] // 3
    for i in range(y*3, y*3 + 3):
        for j in range(x * 3, x*3 + 3):
            if a[i][j] == num and (i,j) != pos:
                return False
    return True


def print_sudoku(a):
    for i in range(len(a)):
        if i % 3 == 0 and i != 0:
            print("- - - - - - - - - - - - - ")
        for j in range(len(a[0])):
            if j % 3 == 0 and j != 0:
                print(" | ", end="")
            if j == 8:
                print(a[i][j])
            else:
                print(str(a[i][j]) + " ", end="")


def find(a):
    for i in range(len(a)):
        for j in range(len(a[0])):
            if a[i][j] == 0:
                return (i, j)  # row, col
    return None
print_sudoku(sudoku)
solve(sudoku)
print("_________________________")
